From f8c919b8f1b60d52879f2e0b2b3f841ad4e04ec9 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 23 May 2024 11:44:35 +0200
Subject: [PATCH 0976/1695] drivers: stm32_cpu_opp: use the supported voltage
 for OPP

Update opp_voltage_is_supported() function to found the first supported
voltage higher that the targeted voltage for the OPP.

This patch simplifies the CPU power supply description on device tree,
and avoids to overdid the OPP description when the CPU power
supply is fixed with voltage not exactly the OPP defined one.

For example:

&cpu0_opp_table {
	opp-650000000 {
		/* Voltage of cpu-supply = <&vddcore> */
		opp-microvolt = <1260000>;
	};
};

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git ad6d5a385ad11340598fd363fa08e9c15ce56a8f drivers: stm32_cpu_opp: use the supported voltage for OPP]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I48c8ee1514bc79763530c6842360df248f4cb076
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/382035
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_cpu_opp.c | 47 +++++++++++++++++++++++++++++++++---
 1 file changed, 43 insertions(+), 4 deletions(-)

diff --git a/core/drivers/stm32_cpu_opp.c b/core/drivers/stm32_cpu_opp.c
index 976efbaaf..084b98d82 100644
--- a/core/drivers/stm32_cpu_opp.c
+++ b/core/drivers/stm32_cpu_opp.c
@@ -102,15 +102,54 @@ static TEE_Result opp_set_voltage(struct regulator *regul, int volt_uv)
 	return regulator_set_voltage(regul, volt_uv);
 }
 
-static bool opp_voltage_is_supported(struct regulator *regul, int volt_uv)
+/*
+ * This function returns true if the given OPP voltage can be managed.
+ * If the exact voltage value is not supported by the regulator,
+ * the function may adjust the input parameter volt_uv to a higher
+ * supported value and still return true.
+ */
+static bool opp_voltage_is_supported(struct regulator *regul, uint32_t *volt_uv)
 {
+	const int target_volt_uv = *volt_uv;
+	int new_volt_uv = 0;
 	int min_uv = 0;
 	int max_uv = 0;
+	unsigned int i = 0;
+	struct regulator_voltages_desc *desc = NULL;
+	const int *levels = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
 
-	regulator_get_range(regul, &min_uv, &max_uv);
+	res = regulator_supported_voltages(regul, &desc, &levels);
+	if (res) {
+		regulator_get_range(regul, &min_uv, &max_uv);
+		if (target_volt_uv > max_uv)
+			return false;
+		if (target_volt_uv < min_uv)
+			*volt_uv = min_uv;
+		return true;
+	}
 
-	if (volt_uv < min_uv || volt_uv > max_uv)
+	if (desc->type == VOLTAGE_TYPE_FULL_LIST) {
+		for (i = 0 ; i < desc->num_levels; i++) {
+			if (levels[i] >= target_volt_uv) {
+				new_volt_uv = levels[i];
+				break;
+			}
+		}
+		if (new_volt_uv == 0)
+			return false;
+
+	} else if (desc->type == VOLTAGE_TYPE_INCREMENT) {
+		new_volt_uv = levels[0]; /* min */
+		while (new_volt_uv < target_volt_uv)
+			new_volt_uv += levels[2]; /* increment */
+		if (new_volt_uv > levels[1]) /* max */
+			return false;
+	} else {
 		return false;
+	}
+
+	*volt_uv = new_volt_uv;
 
 	return true;
 }
@@ -623,7 +662,7 @@ static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
 		}
 
 		/* skip OPP when voltage is not supported */
-		if (!opp_voltage_is_supported(cpu_opp.regul, volt_uv)) {
+		if (!opp_voltage_is_supported(cpu_opp.regul, &volt_uv)) {
 			DMSG("Skip OPP %"PRIu64"kHz/%"PRIu32"uV",
 			     freq_khz, volt_uv);
 			cpu_opp.opp_count--;
-- 
2.47.2

